<a href="https://colab.research.google.com/github/kaledai069/Crossword-Solver-A-Neural-Transformer-based-Approach/blob/master/Building_a_large_word_corpus.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install pyspellchecker

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.5/6.5 MB 13.2 MB/s eta 0:00:00


In [51]:
import nltk
import string

from tqdm import tqdm
from pprint import pprint

In [3]:
nltk.download('wordnet')
nltk.download('words')
nltk.download('abc')
nltk.download('genesis')
nltk.download('gutenberg')
nltk.download('inaugural')
nltk.download('state_union')
nltk.download('webtext')
nltk.download('cmudict')
nltk.download('brown')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package words to /root/nltk_data...
[nltk_data]   Unzipping corpora/words.zip.
[nltk_data] Downloading package abc to /root/nltk_data...
[nltk_data]   Unzipping corpora/abc.zip.
[nltk_data] Downloading package genesis to /root/nltk_data...
[nltk_data]   Unzipping corpora/genesis.zip.
[nltk_data] Downloading package gutenberg to /root/nltk_data...
[nltk_data]   Unzipping corpora/gutenberg.zip.
[nltk_data] Downloading package inaugural to /root/nltk_data...
[nltk_data]   Unzipping corpora/inaugural.zip.
[nltk_data] Downloading package state_union to /root/nltk_data...
[nltk_data]   Unzipping corpora/state_union.zip.
[nltk_data] Downloading package webtext to /root/nltk_data...
[nltk_data]   Unzipping corpora/webtext.zip.
[nltk_data] Downloading package cmudict to /root/nltk_data...
[nltk_data]   Unzipping corpora/cmudict.zip.


True

In [46]:
from nltk.corpus import inaugural, wordnet, words, abc, genesis, gutenberg, state_union, webtext, cmudict, brown

all_corporas = [inaugural, wordnet, words, abc, genesis, gutenberg, state_union, webtext, cmudict, brown]
overall_word_list = []
for corpora in all_corporas:
  corpora_name = str(corpora).split('corpora/')[-1].split("'")[0]
  total_words = len(set(corpora.words()))
  print(f"CORPORA: {corpora_name} WORDS COUNT: {total_words}")

  overall_word_list.extend(list(set(corpora.words())))

print("Total Words: ", len(overall_word_list))

CORPORA: inaugural WORDS COUNT: 10025
CORPORA: wordnet.zip/wordnet/ WORDS COUNT: 147306
CORPORA: words WORDS COUNT: 235892
CORPORA: abc WORDS COUNT: 31885
CORPORA: genesis WORDS COUNT: 25841
CORPORA: gutenberg WORDS COUNT: 51156
CORPORA: state_union WORDS COUNT: 14591
CORPORA: webtext WORDS COUNT: 21538
CORPORA: cmudict WORDS COUNT: 123455
CORPORA: brown WORDS COUNT: 56057
Total Words:  717746


In [47]:
print("After merging all the plaintext corpus:", len(set(overall_word_list)))

After merging all the plaintext corpus: 492435


#### All unique words that can be collected from the nltk.corpus

In [48]:
collected_unique_words = list(set(overall_word_list))

#### Assessing the BCS build words_alph.txt

In [56]:
dictionary = list(set([a.strip() for a in open('/content/words_alpha.txt','r').readlines()]))

for word in tqdm(dictionary, ncols = 100):
  for char in word.lower():
    if char not in string.ascii_lowercase:
      print(word)

100%|███████████████████████████████████████████████████| 370103/370103 [00:03<00:00, 115386.12it/s]


So, there are no words in the dictionary with hypen and any other characters

In [76]:
unique_valid_words = []
invalid_words = []
invalid_char_found = False

for word in tqdm(collected_unique_words, ncols = 120):
  for char in word.lower():
    if char not in string.ascii_lowercase:
      invalid_char_found = True
      break
  if invalid_char_found:
    invalid_char_found = False
    invalid_words.append(word.lower())
  else:
    unique_valid_words.append(word.lower())


100%|███████████████████████████████████████████████████████████████████████| 492435/492435 [00:02<00:00, 203805.59it/s]


In [77]:
filtered_words = []

for word in invalid_words:
  modified_sentence = word.replace('-', ' ').replace('_', ' ').replace("'", '')
  possible_words = modified_sentence.split(' ')

  invalid_char_found_1 = False
  for w in possible_words:
    for char in w.lower():
      if char not in string.ascii_lowercase:
        invalid_char_found_1 = True
        break

    if invalid_char_found_1:
      invalid_char_found_1 = False
    else:
      filtered_words.append(w.lower())

filtered_words = list(set(filtered_words))
len(filtered_words)

42264

In [79]:
collected_valid_words = list(set(unique_valid_words + filtered_words))

print("Unique words length: ", len(unique_valid_words))
print("Words from invalid list length: ", len(filtered_words))
print("All collected unique valid words: ", len(collected_valid_words))
print("BCS words dictionary lenght: ", len(dictionary))

Unique words length:  399224
Words from invalid list length:  42264
All collected unique valid words:  373995
BCS words dictionary lenght:  370103


So, I have successfully collected about 373995 unique valid words, whereas the dictionary build by BCS has 370103 words.

#### Assess the overlap between BCS build dictionary and out built dictionary

In [86]:
own_collected = set(collected_valid_words)
bcs_collected = set(dictionary)

len(own_collected.intersection(bcs_collected))

268623

In [94]:
words_dictionary = own_collected.union(bcs_collected)
sorted_words_dictionary = sorted(list(words_dictionary))
print("Total words in the dictionary: ", len(words_dictionary))

Total words in the dictionary:  475475


In [97]:
output_path = '/content/new_words_alpha.txt'

with open(output_path, 'w') as f:
  f.write("\n".join(list(sorted_words_dictionary)))